setup.cfg

```diff
@@ -72,7 +72,12 @@ all =
     bottleneck
     ipython
     pytest
-docs = sphinx-astropy
+docs =
+    sphinx-astropy
+    pytest
```

Comment:
But... does this mean, `docs` and `all` are mutually exclusive now? Can you still use both, given the duplication?

Reply:
couldn't we reuse on in the other?

Reply:
The original intent to switching over is to avoid specifying a dependency multiple times but seems like we regressed a little. 🤷‍♀

Reply:
yes, that's what bothers me, too. And the difficulty of pinning version numbers for specific builds without pinning it for everything

Reply:
@astrofrog , any thoughts?

Reply:
I recall this came up elsewhere, too when we needed to pin mpl version.

Reply:
Yup, #8827 for RTD

Reply:
Of course.

Reply:
One cannot reuse one in the other if you're going to do a declarative `setup()` using `setup.cfg`.  If you push dependencies back into `setup.py`, then yes, you can define a list for `[docs]` and `[test]` and then combine them (and add a few more to the list) for `[all]`.  I don't think this can be done in `setup.cfg`?

An example for `setup.py`:
```
docs_require = [
    'sphinx-astropy',
    'pytest',
    'pyyaml',
    'scipy',
    'scikit-image',
]
tests_require = [
    'pytest',
    'pytest-astropy',
    'pytest-xdist',
    'pytest-mpl',
    'objgraph',
    'ipython',
    'coverage',
    'skyfield',
]
all_requires = [
    h5py
    beautifulsoup4
    html5lib
    bleach
    pandas
    bintrees
    sortedcontainers
    pytz
    jplephem
    matplotlib>=2.1
    mpmath
    asdf>=2.3
    bottleneck
] + docs_require + tests_require

setup(
    extras_require={
        'docs': docs_require,
        'test': tests_require,
        'all': all_requires,
    },
)
```
Since `pytest` is a dependency of the tests and the docs, it really should be listed in both separately.  Not sure if this added complexity is needed.

Reply:
Yes, that bottleneck has come up earlier. I wish it would work with sets rather than lists, than all this misery would be dealt with easily.

Reply:
Something like
```
all_requires = list(set(docs_require + tests_require))
```
or similar could work.

Of course this will give you a random order of dependencies.  You should probably sort them, as consistent order does matter for reproducibility with `pip install`.